main.go

```diff
@@ -241,22 +209,112 @@ func main() {
+		}, []string{"result"}),
+		queryResponses: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Name: "up_queries_total",
+			Help: "The total number of queries made.",
+		}, []string{"result"}),
+		metricValueDifference: prometheus.NewHistogram(prometheus.HistogramOpts{
+			Name:    "up_metric_value_difference",
+			Help:    "The time difference between the current timestamp and the timestamp in the metrics value.",
+			Buckets: prometheus.LinearBuckets(4, 0.25, 16),
+		}),
+	}
+	reg.MustRegister(
+		prometheus.NewGoCollector(),
+		prometheus.NewProcessCollector(prometheus.ProcessCollectorOpts{}),
+		m.remoteWriteRequests,
+		m.queryResponses,
+		m.metricValueDifference,
+	)
+
+	return m
+}
+
+func parseOptions() (options, error) {
+	var (
+		rawEndpointWrite string
+		rawReadAddress   string
+	)
+
+	opts := options{}
+
+	flag.StringVar(&rawEndpointWrite, "endpoint-write", "", "The endpoint to which to make remote-write requests.")
+	flag.StringVar(&rawReadAddress, "read-address", "",
+		"The base address to which to make query requests to. (/api/v1/query* will be appended to the given address)")
+	flag.Var(&opts.Labels, "labels", "The labels additionally to `__name__` that should be applied to remote-write requests.")
+	flag.StringVar(&opts.Listen, "listen", ":8080", "The address on which internal server runs.")
+	flag.StringVar(&opts.Name, "name", "up", "The name of the metric to send in remote-write requests.")
+	flag.StringVar(&opts.Token, "token", "", "The bearer token to set in the authorization header on remote-write requests.")
+	flag.DurationVar(&opts.Period, "period", 5*time.Second, "The time to wait between remote-write requests.")
+	flag.DurationVar(&opts.Duration, "duration", 5*time.Minute, "The duration of the up command to run until it stops.")
+	flag.Float64Var(&opts.SuccessThreshold, "threshold", 0.9, "The percentage of successful requests needed to succeed overall. 0 - 1.")
+	flag.DurationVar(&opts.Latency, "latency", 15*time.Second, "The maximum allowable latency between writing and reading.")
+	flag.DurationVar(&opts.InitialQueryDelay, "initial-query-delay", 5*time.Second, "The time to wait before executing the first query.")
+	flag.Parse()
+
+	endpointWrite, err := url.ParseRequestURI(rawEndpointWrite)
+	if err != nil {
+		return opts, fmt.Errorf("--endpoint-write is invalid: %w", err)
+	}
+
+	opts.WriteEndpoint = endpointWrite
+
+	var readAddress *url.URL
+	if rawReadAddress != "" {
+		readAddress, err = url.ParseRequestURI(rawReadAddress)
+		if err != nil {
+			return opts, fmt.Errorf("--read-address is invalid: %w", err)
+		}
+	}
+
+	opts.ReadAddress = readAddress
+
+	if opts.Duration <= opts.Period {
+		return opts, errors.New("--duration cannot be less than period")
+	}
+
+	if opts.Latency <= opts.Period {
+		return opts, errors.New("--latency cannot be less than period")
+	}
+
+	if opts.InitialQueryDelay <= opts.Period {
+		return opts, errors.New("--initial-query-delay cannot be less than period")
```

Comment:
why not? you start writing immediately and then every period. You read after the initial delay and then after every period. i want to write every 30 seconds, but only want to wait an initial 5 seconds. Should be legal IMO. These seem like independent things to me.

Reply:
I introduced this because it kind of messes up the reporting metrics if a success threshold introduced. We report higher query failures since we start querying before even write. We can document this and give this flexibility to user.

Reply:
hmm ok but then we have a mistake in the logic. Either there should be no initial delay when writing, or the delay in reading should only be considered after the first write succeeded.

Reply:
e.g. before the read loop: `<-time.After(opts.InitialQueryDelay+opts.Period)`

Reply:
actually, I don't see why setting the initial delay to less than the period would every cause a read before the write.